packages/util-waiter/src/poller.ts

```diff
@@ -18,19 +20,31 @@ const randomInRange = (min: number, max: number) => min + Math.random() * (max -
  * @param stateChecker function that checks the acceptor states on each poll.
  */
 export const runPolling = async <T, S>(
-  { minDelay, maxDelay }: WaiterOptions,
+  { minDelay, maxDelay, maxWaitTime }: ResolvedWaiterOptions,
   client: T,
   input: S,
   acceptorChecks: (client: T, input: S) => Promise<WaiterResult>
 ): Promise<WaiterResult> => {
   let currentAttempt = 1;
-
+  let totalDelay = 0;
+  // The max attempt number that the derived delay time tend to increase.
+  // Pre-compute this number to avoid Number type overflow.
+  const attemptCeiling = Math.log(maxDelay / minDelay) / Math.log(2) + 1;
   while (true) {
-    await sleep(exponentialBackoffWithJitter(minDelay, maxDelay, currentAttempt));
+    const delay = exponentialBackoffWithJitter(minDelay, maxDelay, attemptCeiling, currentAttempt);
+    await sleep(delay);
     const { state } = await acceptorChecks(client, input);
-    if (state === WaiterState.SUCCESS || state === WaiterState.FAILURE) {
+    if (state !== WaiterState.RETRY) {
       return { state };
     }
+
+    totalDelay += delay;
+    if (totalDelay > maxWaitTime) {
```

Comment:
This doesn't take into account the amount of time spent waiting on API calls. Is there a way to just use the actual time and check if the actual time exceeds the deadline?

Reply:
Yea we can. It's a good catch!

Reply:
Fixed in https://github.com/aws/aws-sdk-js-v3/pull/1759/commits/979186a38bcb570f9a6e12abf830eb76caa4714c